apis/ql.d.ts

```diff
@@ -231,29 +231,32 @@ export class DELETE<T> extends ConstructedQuery<T> {
   DELETE: CQN.DELETE['DELETE']
 
 }
+// operator for qbe expression
+type QbeOp = '=' | '-=' | '+=' | '*=' | '/=' | '%='
 
 export interface UPDATE<T> extends Where<T>, And, ByKey {}
 export class UPDATE<T> extends ConstructedQuery<T> {
   private constructor();
 
   // cds-typer plural
-  // FIXME: this returned UPDATE<SingularInstanceType<T>> before. should UPDATE<Books>.entity(...) return Book or Books?
-  static entity<T extends ArrayConstructable> (entity: T, primaryKey?: PK): UPDATE<InstanceType<T>>
-
-  static entity (entity: EntityDescription, primaryKey?: PK): UPDATE<StaticAny>
-
-  static entity<T extends Constructable> (entity: T, primaryKey?: PK): UPDATE<T>
-
-  // currently no easy way to restrict T from being a primitive type
-  static entity<T> (entity: T, primaryKey?: PK): UPDATE<T>
+  static entity: (TaggedTemplateQueryPart<UPDATE<StaticAny>>)
+   // FIXME: this returned UPDATE<SingularInstanceType<T>> before. should UPDATE<Books>.entity(...) return Book or Books?
+    & (<T extends ArrayConstructable> (entity: T, primaryKey?: PK) => UPDATE<SingularInstanceType<T>>)
+    & (<T extends Constructable> (entity: T, primaryKey?: PK) => UPDATE<InstanceType<T>>)
+    & ((entity: EntityDescription, primaryKey?: PK) => UPDATE<StaticAny>)
+    & (<T> (entity: T, primaryKey?: PK) => UPDATE<T>)
 
   // with (block: (e:T)=>void) : this
   // set (block: (e:T)=>void) : this
-  set: TaggedTemplateQueryPart<this>
-  & ((data: object) => this);
 
+  // simple value   > title: 'Some Title'
+  // qbe expression > stock: { '-=': quantity }  
+  // cqn expression > descr: {xpr: [{ref:[descr]}, '||', 'Some addition to descr.']}
+  set: TaggedTemplateQueryPart<this>
+    & ((data: {[P in keyof T]?: T[P] | {[op in QbeOp]?: any} | CQN.xpr}) => this)
+    
   with: TaggedTemplateQueryPart<this>
-    & ((data: object) => this)
+    & ((data: {[P in keyof T]?: T[P] | {[op in QbeOp]?: any} | CQN.xpr}) => this)
```

Comment:
I guess this is the part which you said you couldn't get to work? If so, then I suggest:

```suggestion
    & ((data: {[P in keyof T]?: T[P] | {[op in QbeOp]?: _TODO} | CQN.xpr}) => this)
```

Reply:
Same for the type for `set`, obviously. :)
For now, this is probably fine to duplicate. But as soon as we introduce one more overload, this should be DRYd out into a separate type.

Reply:
Should be possible with this:

```ts
with: typeof this.set
```

Reply:
that would be preferable, yes. 👍

Reply:
But the `any` here is not the part I was mentioning in the PR. I wasn't sure if  `{[op in QbeOp]?: T[P]}` would be ok or not. 

If we take the type Foo

```cds
entity Foo {
  x: String
}
```

then the following query:

```ts
UPDATE(Foos).set({ x: {  '+=': 5 } });
```

will show an error because `number` does not match the type `string`. If this is ok, then we can go for `T[P]` instead of `any`

Reply:
Is it a valid use case for a user to have a string property and try to query by a number example? I guess they could have an entry with `x = '5'` and hope for the system to coerce their numeric example to the proper type. My gut feeling is to agree with your assumption to go for `T[P]`. But I don't have much of a user's perspective, so that's something you know more about than I do. 🙂

Reply:
```ts
function fn (p: string) {}
fn (42)  // Argument of type 'number' is not assignable to parameter of type 'string'.
```
so when you are a TS user, operating with a TS mind set, then coercion is usually not a thing wrt parameters. I'd be forced to explicitly stringify the numeric parameter.

Reply:
Unfortunately I don't use the qbe or cqn-expression variants much or at all. I just included them in addition to the simple variant `{ x : 5 }` for completion sake 😊

Reply:
I personally would either risk it and go with `T[P]` or we consult the runtime team to make sure

Reply:
Given the last sync I had with the runtime, where we opted to be as strict as possible whenever we do have a type at hand, I'd say: go for `T[P]`! 😉